# LAB 02.03 - Pandas

In [1]:
!wget --no-cache -O init.py -q https://raw.githubusercontent.com/rramosp/ai4eng.v1/main/content/init.py
import init; init.init(force_download=False); init.get_weblink()


replicating local resources


In [2]:
from local.lib.rlxmoocapi import submit, session
session.LoginSequence(endpoint=init.endpoint, course_id=init.course_id, lab_id="L02.03", varname="student");

logging in as andres.lema1@udea.edu.co... please wait

-------------
using course session ai4eng.v1.udea.r3::2025.2
success!! you are logged in
-------------


In [3]:
import numpy as np
import pandas as pd

## Task 1: Extract data

we have a dataframe of items with a price and, **sometimes**, with an extra column (`margin`).

You will have to complete a function that will **filter** the dataframe selecting the rows:

- whose price is > 100
- **OR** whose margin is >10, if the column `margin` is present

your function must return **A LIST** with the item ids of the selected rows

For instance, with the following dataframe

             price  category
    itemid                  
    56556   108.15       2.0
    73065    83.54       2.0
    36619   114.42       0.0
    73414    82.94       0.0
    13410   115.13       2.0
    66153    91.50       1.0
    77380    85.82       0.0
    73249    95.95       0.0
    11654   100.22       2.0
    11972    77.16       1.0

your must return the following list:

     [56556, 36619, 13410, 11654]
     
But if you get the following dataframe

             price  category  margin
    itemid                          
    39059    98.11       0.0   11.04
    19526    98.11       1.0   11.25
    78176    94.34       1.0   10.51
    50948   102.37       1.0   10.77
    12111    98.07       1.0    8.50
    56191    98.53       1.0   11.65
    38887    91.49       2.0   11.24
    77915   117.30       0.0    8.64
    55010    96.13       0.0    8.95
    45925    98.59       1.0   10.45
    
you must return the following list

    [39059, 19526, 78176, 50948, 56191, 38887, 77915, 45925]

In [4]:
def create_df(missing=False, n=10):
    itemid   = np.random.randint(100000, size=n)+1000
    category = np.random.randint(3, size=n)
    price    = np.round(np.random.normal(loc=100, scale=10, size=n),2)
    margin   = np.round(np.random.normal(loc=10, scale=1, size=n),2)

    if missing:
        nmissing = np.random.randint(len(price)//2)+2
        price[np.random.permutation(len(price))[:nmissing]] = np.nan

    d = pd.DataFrame(np.r_[[price, category, margin]].T, index=itemid, columns=["price", "category", "margin"])
    d.index.name="itemid"
    if np.random.random()>.5:
        d = d[d.columns[:2]]

    return d

In [5]:
d = create_df()
d

price  category
itemid                  
26204    94.81       0.0
75675   107.84       0.0
77453    97.85       2.0
35988    90.65       1.0
95484   101.21       0.0
76764    87.63       2.0
94793    95.84       1.0
52686   112.96       1.0
10162   100.91       1.0
92139   101.35       0.0

In [6]:
def select_items(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()

    if "margin" in df.columns:
        mask = (df["price"] > 100) | (df["margin"] > 10)
    else:
        mask = df["price"] > 100

    result = df.index[mask].tolist()

    return result

**manually check your answer**

In [7]:
d = create_df()
d

price  category
itemid                  
38644    93.21       2.0
14506   104.03       1.0
7870     94.48       2.0
33621   108.86       1.0
69143   100.76       1.0
85359    96.05       1.0
75159   104.84       2.0
47576    93.56       1.0
7105    110.99       2.0
94782    92.77       2.0

In [8]:
select_items(d)

[14506, 33621, 69143, 75159, 7105]

**submit your code**

In [9]:
student.submit_task(globals(), task_id="task_01");

## Task 2: Group statistics

Complete the following function so that it returns a dataframe with the average, max and min **prices** per category.

For instance, for the following dataframe

             price  category  margin
    itemid                          
    17946    93.85       1.0   10.64
    61190    91.72       1.0    9.76
    39639   100.16       1.0   10.67
    17791   110.44       2.0    9.65
    7333    101.05       1.0    9.69
    77362   122.33       0.0   11.14
    92646   108.13       2.0   10.58
    27797    85.52       2.0   10.88
    31746    97.56       0.0    9.75
    12355   101.04       2.0    9.51
    
you should return the following dataframe

                 media  maximo  minimo
    categoria                         
    0         109.9450  122.33  97.56
    1          96.6950  101.05  91.72
    2         101.2825  110.44  85.52
    
observe that your result
- must not be a **multilevel** columnset.
- the column names and the index name must be **exactly** as in the example.
- the **index** must be of type **int**.

In [10]:
def get_stats(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()

    result = df.groupby("category")["price"].agg(
        media="mean",
        maximo="max",
        minimo="min"
    )

    result.index = result.index.astype(int)
    result.index.name = "categoria"

    return result

**manually check your answer**

In [11]:
d = create_df()
d

price  category  margin
itemid                          
84986    89.99       1.0    9.21
34582   110.32       2.0   11.54
94330    90.86       1.0    9.10
44828    89.10       2.0   10.30
65021    82.25       0.0    8.62
29322   101.43       1.0    9.85
42988   104.70       0.0    9.34
55735   114.49       2.0    9.53
56280   106.44       1.0   11.06
73131    94.01       2.0    9.82

In [12]:
get_stats(d)

media  maximo  minimo
categoria                         
0           93.475  104.70   82.25
1           97.180  106.44   89.99
2          101.980  114.49   89.10

**submit your code**

In [13]:
student.submit_task(globals(), task_id="task_02");

## Task 3: Fill in missing data

Fill in the missing data in the **price** column with the following procedure:

- compute the mean and std of the avaialable prices
- sample from a normal distribution with the computed mean and std (see [`np.random.normal`])(https://numpy.org/doc/stable/reference/random/generated/numpy.random.normal.html), as many samples as missing data
- substitute the missing values with the samples

For instance, for this input dataframe:

             price  category  margin
    itemid                          
    18922      NaN       1.0   10.32
    69500   121.25       1.0   10.22
    76442    90.25       1.0   12.60
    33863   106.51       0.0   10.26
    15904    95.87       1.0   11.51
    41946   103.47       2.0    9.85
    85451    93.08       2.0    9.56
    70028   116.68       1.0    9.11
    26860      NaN       2.0    9.71
    12807    91.48       0.0    9.77
    
your solution might be similar to this (not exactly the same as you will be sampling data):


                 price  category  margin
    itemid                              
    18922    97.441188       1.0   10.32
    69500   121.250000       1.0   10.22
    76442    90.250000       1.0   12.60
    33863   106.510000       0.0   10.26
    15904    95.870000       1.0   11.51
    41946   103.470000       2.0    9.85
    85451    93.080000       2.0    9.56
    70028   116.680000       1.0    9.11
    26860   103.294843       2.0    9.71
    12807    91.480000       0.0    9.77
    
    
**WARN**: your function must not modify the original dataframe, make a copy of the input dataframe, fill the values in the copy and return it.

**HINT**: use the [`isna()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isna.html) method of a dataframe or a series.

In [14]:
d = create_df(missing=True)
d

price  category
itemid                  
66855    91.55       1.0
50551      NaN       1.0
59699    93.54       1.0
88669      NaN       2.0
43193      NaN       0.0
37188   103.60       2.0
66975      NaN       0.0
59408    79.88       0.0
50740    96.61       0.0
30707   116.93       2.0

In [15]:
def fillna(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()

    # YOUR CODE HERE
    mean = df["price"].mean(skipna=True)
    std = df["price"].std(skipna=True)

    missing_mask = df["price"].isna()
    n_missing = missing_mask.sum()

    samples = np.random.normal(mean, std, n_missing)

    df.loc[missing_mask, "price"] = samples

    result = df

    return result

**check manually your code**

In [16]:
d = create_df(missing=True)
d

price  category
itemid                  
66613      NaN       0.0
96700   115.01       1.0
65815   107.82       0.0
90356      NaN       1.0
2180       NaN       2.0
92923   100.18       2.0
61377    98.57       0.0
18196   111.33       1.0
96018    93.12       2.0
34490   111.65       1.0

In [17]:
fillna(d)

price  category
itemid                      
66613   111.624606       0.0
96700   115.010000       1.0
65815   107.820000       0.0
90356   113.137013       1.0
2180    110.064393       2.0
92923   100.180000       2.0
61377    98.570000       0.0
18196   111.330000       1.0
96018    93.120000       2.0
34490   111.650000       1.0

**submit your code**

In [18]:
student.submit_task(globals(), task_id="task_03");